### ISSUE TITLE
Duck typed dask collections for use with dask.imperative

### ISSUE THREAD

**Issue by shoyer** [dask/dask#700]
If an `xray.Dataset` consists of dask arrays, it should work like a first class dask collection for integration with `dask.imperative`. That is, when you call a delayed function (constructed with `dask.imperative.do`) on a xray Dataset consisting of dask arrays, dask should merge the task graphs when it executes.

The logic in dask.imperative that does this currently checks if the object inherits from the dask collection base class: 
https://github.com/blaze/dask/blob/7e9bff047894c2bc9539370898dc135508739d38/dask/imperative.py#L67-L72

Strictly speaking, xray objects aren't always dask collections (dask isn't even a required dependency), so some type of duck typing solution seems appropriate. What should this look like?

The simplest way to adapt existing code would be to check for `dask`, `_keys`, `_optimize` and `_finalize` attributes, but I would prefer more specific names for the private methods to make it clearer what they refer to. I'm happy with a `.dask` attribute (the presence of which alone might be enough to signal a dask collection), but for the others, maybe  `__dask_keys__`, `__dask_optimize__` and `__dask_finalize__` would be appropriate names?

**Comment by jcrist** [comment]
Proposed interface:

```python
class Base(object):
    def _dask_optimize_(self, dsk, keys, **kwargs):
        """Return an optimized dask graph"""
        pass

    def _dask_finalize_(self, results):
        """Finalize the computed keys into an output object"""
        pass

    def _dask_default_get_(self, dsk, key, **kwargs):
        """The default scheduler `get` to use for this collection"""
        pass

    def _dask_keys_(self):
        """The keys for the dask graph"""
        pass

    @property
    def _dask_graph_(self):
        """The graph"""
        pass
```

Note that in most cases these will all be attributes/staticmethods.

**Comment by mrocklin** [comment]
In cases like finalize they actually have to be staticmethods.  We can't put any instance methods of the original collection into the dask graph itself.

**Comment by mrocklin** [comment]
This is important for distributed computing.

**Comment by jcrist** [comment]
Not necessarily, they could be properties that return a function. For example for xarray, I'd do something like:

```python
@property
def _dask_finalize_(self):
    return partial(xarray_finalize, dims=self.dims, coords=self.coords, name=self.name, attrs=self.attrs)
```

This avoids expensive serialization of `self`, but allows enclosing the needed metadata.

**Comment by mrocklin** [comment]
OK, that would work for me.

**Comment by shoyer** [comment]
This looks like a pretty sane API to me. Just to be clear, only `_dask_finalize_` needs to be a property that returns a function?

@jcrist's version of `_dask_finalize_` for xarray is pretty close but maybe not quite right, because we can also have dask arrays in coordinates as well as the data itself. Although, it's not entirely clear if calling `compute()` on such an array should compute the coordinates, too...

**Comment by jcrist** [comment]
Because the interface methods are only accessed as attributes on the objects, they could be anything (method, staticmethod, attribute, property, etc...). The reason for avoiding methods for `_dask_finalize_` is to avoid unnecessary and costly serialization of the object itself when used with `dask.imperative` (the finalize function gets added to the graph). Defining it as a property above gets around that, while still keeping relative metadata (note that that was just an example, and not the actual implementation).

**Comment by jcrist** [comment]
This was fixed in #2748. Closing.
